File: xpilot/contrib/jxpmap/python/MapPolygon.py
from javastuff import *
from MapObject import MapObject
from PolygonStyle import PolygonStyle
from LineStyle import LineStyle
from PolygonPropertyEditor import PolygonPropertyEditor

import gtk
import GDK

class MapPolygon(MapObject):
    TYPE_NORMAL = 0
    TYPE_BALLAREA = 1
    TYPE_BALLTARGET = 2
    TYPE_DECORATION = 3

    def __init__(self, p = None, style = None, edgeStyles = None,bmstore=None):
        self.polygon = p
        self.style = style
        self.edgeStyles = edgeStyles

    def getCreateHandler(self):
        return CreateHandler(self)

    def setType(self, type):
        self.type = type

    def getType(self):
        return self.type

    def setTeam(self, team):
        self.team = team

    def getBounds(self):
        return self.polygon.getBounds()

    def contains(self, canvas, p):
        for wrap in canvas.wrapOffsets(self.getBounds(), p):
            if self.polygon.contains(Point(p.x + wrap[0], p.y + wrap[1])):
                return 1
        return 0

    def getZOrder(self):
        return 20

    def getPreviewShape(self):
        return self.polygon

    def moveTo(self, x, y):
        r = self.getBounds()
        self.polygon.transform(AffineTransform(x - r.x, y - r.y))

    def paint(self, di):
        gc = di.gc
        colormap = di.colormap
        tx = di.tx

        trans = [tx(p) for p in self.polygon.points]
        points = [(t.x, t.y) for t in trans]

        if self.style.getFillStyle() == PolygonStyle.FILL_COLOR:
            gc.foreground = colormap.alloc(*gcolor(self.style.getColor()))
            gtk.draw_polygon(di.area, gc, gtk.TRUE, points)
        elif self.style.getFillStyle() == PolygonStyle.FILL_TEXTURED:
            raise "unimplemented"

        if self.edgeStyles is None:
            df = self.style.getDefaultEdgeStyle()
            if df.getStyle() != LineStyle.STYLE_HIDDEN:
                gc.foreground = colormap.alloc(*gcolor(df.getColor()))
                gtk.draw_polygon(di.area, gc, gtk.FALSE, points)
                # !@# dashed lines, width missing here
        else:
            raise "unimplemented"

    def getPropertyEditor(self, canvas):
        return PolygonPropertyEditor(canvas, self)

    def checkEvent(self, canvas, me):
        b = self.getBounds()
        p = me.getPoint()
        pl = self.polygon
        THRESHOLD = 25 / canvas.scale**2

        larger = Rectangle(b.x - 20 * 64, b.y - 20 * 64,
                           b.width + 40 * 64, b.height + 40 * 64)
        wraps = canvas.wrapOffsets(larger, p)
        if len(wraps) == 0:
            return 0

        if me.button == 1:
            for i in range(len(pl.points)):
                point = pl.points[i]
                if canvas.wrapdist2(p, point) < THRESHOLD:
                    if canvas.isErase():
                        if self.polygon.npoints > 3:
                            self.removePoint(i)
                            canvas.repaint()
                    else:
                        canvas.setCanvasEventHandler(PolygonPointMoveHandler(
                            self, me, i, canvas.wrap(p, point)))
                    return 1

        if not canvas.isErase():
            for wrap in wraps:
                for i in range(pl.npoints):
                    p1 = pl.points[i - 1]
                    p2 = pl.points[i]
                    if ptSeqDistSq(p1.x, p1.y, p2.x, p2.y, p.x + wrap[0],
                                   p.y + wrap[1]) < THRESHOLD:
                        if me.button == 1:
                            self.insertPoint(i, p)
                            canvas.setCanvasEventHandler(PolygonPointMoveHandler(
                                self, me, i, wrap))
                            return 1

        return MapObject.checkEvent(self, canvas, me)

    def removePoint(self, i):
        del self.polygon.points[i]
        self.polygon.npoints -= 1
        self.polygon.recalculate_bounds()
        if self.edgeStyles is not None:
            del self.edgeStyles[i - 1] # works right for i == 0

    def insertPoint(self, i, p):
        self.polygon.points.insert(i, p)
        self.polygon.npoints += 1
        self.polygon.recalculate_bounds()
        if self.edgeStyles is not None:
            self.edgeStyles.insert(i - 1, self.edgeStyles[i - 1])

    def isCounterClockwise(self):
        x1 = long(self.polygon.points[-1].x)
        y1 = long(self.polygon.points[-1].y)
        area = 0
        for p in self.polygon.points:
            x2 = long(p.x)
            y2 = long(p.y)
            area += x1 * y2 - x2 * y1
            x1, y1 = x2, y2
        return area > 0

    def printXml(self, file):
        if self.polygon.npoints < 3:
            raise "Invalid polygon with less than 3 points"

        if self.getType() == self.TYPE_BALLAREA:
            file.write('<BallArea>\n')
        elif self.getType() == self.TYPE_BALLTARGET:
            file.write('<BallTarget team="%d">\n' % self.team)
        elif self.getType() == self.TYPE_DECORATION:
            file.write('<Decor>\n')

        points = self.polygon.points[:]
        es = self.edgeStyles
        cls = None
        if es is None:
            es = [None] * len(points)
        file.write('<Polygon x="%d" y="%d" style="%s">\n' % (
                   points[0].x, points[0].y, self.style.getId()))

        if not self.isCounterClockwise():
            points.reverse()
            es = es[:]
            es.reverse()
        else:
            points.append(points.pop(0))

        pp = points[-1]
        for p, e in zip(points, es):
            file.write('<Offset x="%d" y="%d"' % (p.x - pp.x, p.y - pp.y))
            if e is not cls:
                if e is None:
                    id = self.style.getDefaultEdgeStyle().getId()
                else:
                    id = e.getId()
                file.write(' style="%s"' % e)
            file.write('/>\n')
            pp = p
        file.write('</Polygon>\n')

        if self.getType() == self.TYPE_BALLAREA:
            file.write('</BallArea>\n')
        elif self.getType() == self.TYPE_BALLTARGET:
            file.write('</BallTarget>\n')
        elif self.getType() == self.TYPE_DECORATION:
            file.write('</Decor>\n')


# !@# Can create starting points with incorrect wrapping
class CreateHandler(MouseEventHandler):
    def __init__(self, object):
        self.object = object
        self.poly = None
        self.remove = 0

    def mousePressed(self, me):
        c = me.getSource()
        gc = c.previewgc
        if self.poly is None:
            self.poly = Polygon()
            first = me.getPoint()
            self.poly.addPoint(first.x, first.y)
            self.poly.addPoint(first.x, first.y)
            return
        c = me.getSource()
        if self.remove:
            c.drawShape(self.poly, gc)
            self.remove = 0
        latest = me.getPoint()
        self.poly.points[self.poly.npoints - 1] = latest
        if me.button == 3:
            self.poly.recalculate_bounds()
            self.object.polygon = self.poly
            self.object.style = c.getModel().getDefaultPolygonStyle()
            c.getModel().addToFront(self.object)
            c.setCanvasEventHandler(None)
            c.repaint()
            return
        self.poly.addPoint(latest.x, latest.y)
        c.drawShape(self.poly, gc)
        self.remove = 1

    def mouseMoved(self, me):
        c = me.getSource()
        if self.poly is None:
            return
        if self.remove:
            c.drawShape(self.poly)
            self.remove = 0
        self.poly.points[self.poly.npoints - 1] = me.getPoint()
        # Bounds not recalculated here so might not be 100% correct
        c.drawShape(self.poly)
        self.remove = 1

class PolygonPointMoveHandler(MouseEventHandler):
    def __init__(self, obj, me, index, wrap):
        self.obj = obj
        self.index = index
        self.virgin = 1
        self.wrap = wrap
        ip = index - 1 % obj.polygon.npoints
        prev = obj.polygon.points[(index - 1) % obj.polygon.npoints]
        self.l1 = Line(prev, prev)
        next = obj.polygon.points[(index + 1) % obj.polygon.npoints]
        self.l2 = Line(next, next)

    def mouseMoved(self, evt):
        c = evt.getSource()
        if not self.virgin:
            self.drawPreview(c)
        else:
            self.virgin = 0
        p = evt.getPoint()
        p.x += self.wrap[0]
        p.y += self.wrap[1]
        self.l1.p2 = self.l2.p2 = evt.getPoint()
        # Bounds not recalculated...
        self.drawPreview(c)

    def mouseReleased(self, evt):
        c = evt.getSource()
        p = evt.getPoint()
        p.x += self.wrap[0]
        p.y += self.wrap[1]
        self.obj.polygon.points[self.index] = evt.getPoint()
        self.obj.polygon.recalculate_bounds()
        c.setCanvasEventHandler(None)
        c.repaint()

    def drawPreview(self, c):
        c.drawShape(self.l1)
        c.drawShape(self.l2)

File: xpilot/contrib/jxpmap/org/xpilot/jxpmap/MapPolygon.java
package org.xpilot.jxpmap;

import java.awt.*;
import java.awt.event.*;
import java.awt.geom.*;
import java.awt.image.BufferedImage;
import java.util.List;
import java.util.ArrayList;

import java.io.PrintWriter;
import java.io.IOException;


public class MapPolygon extends MapObject {

    public static final int TYPE_NORMAL = 0;
    public static final int TYPE_BALLAREA = 1;
    public static final int TYPE_BALLTARGET = 2;


    protected Polygon polygon;
    protected PolygonStyle style;
    protected ArrayList edgeStyles;
    protected int type;
    protected int team;


    public MapPolygon () {}


    public MapPolygon (Polygon p, PolygonStyle style, ArrayList edgeStyles) {
        this.polygon = p;
        this.style = style;
        this.edgeStyles = edgeStyles;
    }

    
    public CanvasEventHandler getCreateHandler (Runnable r) {
        return new CreateHandler(r);
    }


    public PolygonStyle getStyle () {
        return style;
    }


    public void setStyle (PolygonStyle style) {
        this.style = style;
    }


    public int getType () {
        return type;
    }


    public void setType (int type) {
        this.type = type;
    }


    public int getTeam () {
        return team;
    }


    public void setTeam (int team) {
        this.team = team;
    }


    public Rectangle getBounds () {
        return polygon.getBounds();
    }


    public boolean contains (Point p) {
        return polygon.contains(p);
    }


    public int getZOrder () {
        return 20;
    }


    public Shape getPreviewShape () {
        return polygon;
    }


    public void moveTo (int x, int y) {
        Rectangle r = getBounds();
        polygon.translate(x - r.x, y - r.y);
    }

    
    public void printXml (PrintWriter out) throws IOException {

        if (polygon.npoints == 0) return;

        if (getType() == TYPE_BALLAREA) {
            out.print("<BallArea team=\"");
            out.print(getTeam());
            out.println("\">");
        } else if (getType() == TYPE_BALLTARGET) {
            out.print("<BallTarget team=\"");
            out.print(getTeam());
            out.println("\">");
        }

        out.print("<Polygon x=\"");
        out.print(polygon.xpoints[0]);
        out.print("\" y=\"");
        out.print(polygon.ypoints[0]);
        out.print("\" style=\"");
        out.print(getStyle().getId());
        out.println("\">");

        for (int i = 1; i < polygon.npoints; i++) {
            out.print("<Offset x=\"");
            out.print(polygon.xpoints[i] - polygon.xpoints[i - 1]);
            out.print("\" y=\"");
            out.print(polygon.ypoints[i] - polygon.ypoints[i - 1]);
            
            LineStyle es = getEdgeStyle(i - 1);
            if (es != null && es != getStyle().getDefaultEdgeStyle()) {
                out.print("\" style=\"");
                out.print(es.getId());
            }
            out.println("\"/>");
        }

        out.println("</Polygon>");

        if (getType() == TYPE_BALLAREA) {
            out.println("</BallArea>");
        } else if (getType() == TYPE_BALLTARGET) {
            out.println("</BallTarget>");
        }
    }
    

    public void paint (Graphics2D g, float scale) {

        if (!style.isVisible()) return;

        Polygon p = polygon;

        if (style.getFillStyle() == PolygonStyle.FILL_COLOR) {
            g.setColor(style.getColor());
            g.fillPolygon(p);
            
        } else if (style.getFillStyle() == PolygonStyle.FILL_TEXTURED) {
            BufferedImage img = style.getTexture().getImage();
            Rectangle b = polygon.getBounds();
            g.setPaint(new TexturePaint(img, new Rectangle
                (b.x, b.y, img.getWidth() * 64, -img.getHeight() * 64)));
            g.fill(p);
        }

        if (edgeStyles == null) {
            LineStyle def = style.getDefaultEdgeStyle();
            if (def.getStyle() != LineStyle.STYLE_HIDDEN) {
                g.setColor(def.getColor());
                g.setStroke(def.getStroke(scale));
                g.draw(p);
            }
        } else {
            for (int i = 0; i < p.npoints - 1;) {
                int begin = i;
                LineStyle ls = (LineStyle)getEdgeStyle(i++);
                
                while (i < p.npoints - 1 && getEdgeStyle(i) == ls) i++;
                
                if (ls == null) ls = style.getDefaultEdgeStyle();
                
                if (ls.getStyle() != LineStyle.STYLE_HIDDEN) {
                    g.setColor(ls.getColor());
                    g.setStroke(ls.getStroke(scale));
                    
                    int nump = i + 1 - begin;
                    int[] xp = new int[nump];
                    int[] yp = new int[nump];

                    System.arraycopy(p.xpoints, begin, xp, 0, nump);
                    System.arraycopy(p.ypoints, begin, yp, 0, nump);

                    g.drawPolyline(xp, yp, nump);
                }
            }
        }
    }


    public void setEdgeStyle (int index, LineStyle style) {
        if (edgeStyles == null) edgeStyles = new ArrayList();
        while (edgeStyles.size() < index + 1) edgeStyles.add(null);
        edgeStyles.set(index, style);
    }


    public LineStyle getEdgeStyle (int index) {
        return 
            (edgeStyles == null || index > edgeStyles.size() - 1) ?
            style.getDefaultEdgeStyle() : 
            (LineStyle)edgeStyles.get(index);
    }


    public EditorPanel getPropertyEditor (MapCanvas canvas) {
        return new PolygonPropertyEditor(canvas, this);
    }


    public boolean checkAwtEvent (MapCanvas canvas, AWTEvent evt) {

        if (evt.getID() == MouseEvent.MOUSE_PRESSED ||
            evt.getID() == MouseEvent.MOUSE_CLICKED) {

            MouseEvent me = (MouseEvent)evt;
            Rectangle b = getBounds();
            Point p = me.getPoint();
            Polygon pl = polygon;


            if ((me.getModifiers() & InputEvent.BUTTON1_MASK) != 0
                && me.getID() == me.MOUSE_PRESSED) {
                
                if (b.x - 20 * 64 > p.x
                    || b.y - 20 * 64 > p.y
                    || b.x + b.width + 20 * 64 < p.x
                    || b.y + b.height + 20 * 64 < p.y) return false;
                
                for (int i = 0; i < pl.npoints; i++) {
                    if (Point.distanceSq
                        (p.x, p.y, pl.xpoints[i], 
                         pl.ypoints[i]) < 25 * 64 * 64) {
                        
                        if (canvas.isErase()) {
                            int pc = pl.npoints - 1;
                            int[] xps = remove(pl.xpoints, i);
                            int[] yps = remove(pl.ypoints, i);
                            
                            polygon = new Polygon(xps, yps, pc);
                            canvas.repaint();
                            
                        } else {
                            canvas.setCanvasEventHandler
                                (new PolygonPointMoveHandler(me, i));
                        }
                        return true;
                    }
                }
            }
                
            if (!canvas.isErase()) {
                    
                for (int i = 0; i < pl.npoints; i++) {
                        
                    int ni = (i + 1) % pl.npoints;
                    if (Line2D.ptSegDistSq(pl.xpoints[i],
                                           pl.ypoints[i],
                                           pl.xpoints[ni],
                                           pl.ypoints[ni],
                                           p.x, p.y) < 25 * 64 * 64) {

                        if ((me.getModifiers() & 
                             InputEvent.BUTTON1_MASK) != 0) {
                            
                            if (me.getID() == me.MOUSE_PRESSED) {
                            
                                int pc = pl.npoints + 1;
                                int xps[] = insert(p.x, pl.xpoints, i + 1);
                                int yps[] = insert(p.y, pl.ypoints, i + 1);
                                
                                polygon = new Polygon(xps, yps, pc);
                                
                                canvas.setCanvasEventHandler
                                    (new PolygonPointMoveHandler(me, i + 1));
                                return true;
                            }

                        } else {

                            if (me.getID() == me.MOUSE_CLICKED) {
                                EditorDialog.show
                                    (canvas,
                                     new EdgePropertyEditor(canvas, this, i),
                                     true,
                                     EditorDialog.OK_CANCEL);
                                return true;
                            }
                        }
                    }
                }
            }
        }

        return super.checkAwtEvent(canvas, evt);
    }


    private static int[] remove (int a[], int p) {

        if (p >= a.length || p < 0) 
            throw new ArrayIndexOutOfBoundsException(p);

        int rv[] = new int[a.length - 1];
        if (p > 0) System.arraycopy(a, 0, rv, 0, p);
        if (p < rv.length) System.arraycopy(a, p + 1, rv, p, rv.length - p);

        return rv;
    }


    private static int[] insert (int i, int a[], int p) {

        if (p >= a.length || p < 0) 
            throw new ArrayIndexOutOfBoundsException(p);

        int rv[] = new int[a.length + 1];
        System.arraycopy(a, 0, rv, 0, p);
        rv[p] = i;
        if (p < a.length) System.arraycopy(a, p, rv, p + 1, a.length - p);

        return rv;
    }



    private class CreateHandler extends CanvasEventAdapter {

        private Runnable cmd;
        private Polygon poly;
        private Point latest, first;
        private Stroke stroke;
        private boolean remove;


        public CreateHandler (Runnable cmd) {
            this.cmd = cmd;
        }

        
        public void mousePressed (MouseEvent me) {

            if (poly == null) {
                poly = new Polygon();
                first = me.getPoint();
                poly.addPoint(first.x, first.y);
                poly.addPoint(0, 0);
                return;
            }

            MapCanvas c = (MapCanvas)me.getSource();
            Graphics2D g = (Graphics2D)c.getGraphics();
            g.setXORMode(Color.black);
            g.setColor(Color.white);
            if (stroke == null) stroke = getPreviewStroke(c.getScale());
            g.setStroke(stroke);

            if (remove) {
                c.drawShape(g, poly);
                remove = false;
            }

            latest = me.getPoint();

            if (poly.npoints > 2) {
                if (first.distanceSq(latest) < 100 * 64 * 64) {
                    poly.xpoints[poly.npoints - 1] = first.x;
                    poly.ypoints[poly.npoints - 1] = first.y;
                    MapPolygon.this.polygon = poly;
                    MapPolygon.this.style = 
                        c.getModel().getDefaultPolygonStyle();
                    
                    c.getModel().addObject(MapPolygon.this);
                    c.setCanvasEventHandler(null);
                    if (cmd != null) cmd.run();
                    c.repaint();
                    return;
                }
            }
            
            poly.addPoint(latest.x, latest.y);
            c.drawShape(g, poly);
            remove = true;
        }


        public void mouseMoved (MouseEvent me) {
            
            if (poly == null) return;

            MapCanvas c = (MapCanvas)me.getSource();
            Graphics2D g = (Graphics2D)c.getGraphics();
            g.setXORMode(Color.black);
            g.setColor(Color.white);
            if (stroke == null) stroke = getPreviewStroke(c.getScale());
            g.setStroke(stroke);
            if (remove) c.drawShape(g, poly);
            
            int li = poly.npoints - 1;
            poly.xpoints[li] = me.getX();
            poly.ypoints[li] = me.getY();
            
            c.drawShape(g, poly);
            remove = true;
        }
        
    }


    private class PolygonPointMoveHandler extends CanvasEventAdapter {


        private int index;
        private Line2D.Float l1, l2;
        private boolean virgin;
        private Stroke stroke;
        

        public PolygonPointMoveHandler (MouseEvent me, int index) {

            this.index = index;
            virgin = true;

            int ip = index - 1;
            if (ip < 0) ip += polygon.npoints;
            Point prev = new Point(polygon.xpoints[ip], polygon.ypoints[ip]);
            l1 = new Line2D.Float(prev, me.getPoint());
            
            int in = (index + 1) % polygon.npoints;
            Point next = new Point(polygon.xpoints[in], polygon.ypoints[in]);
            l2 = new Line2D.Float(next, me.getPoint());

            
            float dash[] = { 10.0f };
            stroke = new BasicStroke(1, BasicStroke.CAP_BUTT, 
                                     BasicStroke.JOIN_MITER, 
                                     10.0f, dash, 0.0f);
        }


        public void mouseDragged (MouseEvent evt) {

            MapCanvas c = (MapCanvas)evt.getSource();
            Graphics2D g = (Graphics2D)c.getGraphics();
            g.setXORMode(Color.black);
            g.setColor(Color.white);
            g.setStroke(stroke);

            if (!virgin) drawPreview(g, c);
            else virgin = false;

            l1.x2 = l2.x2 = evt.getX();
            l1.y2 = l2.y2 = evt.getY();
            
            drawPreview(g, c);
        }
        

        public void mouseReleased (MouseEvent evt) {
            
            polygon.xpoints[index] = evt.getX();
            polygon.ypoints[index] = evt.getY();
            Polygon p = new Polygon(polygon.xpoints, 
                                    polygon.ypoints, 
                                    polygon.npoints);
            polygon = p;
            MapCanvas c = (MapCanvas)evt.getSource();
            c.setCanvasEventHandler(null);
            c.repaint();
        }


        private void drawPreview (Graphics2D g, MapCanvas c) {
            c.drawShape(g, l1);
            c.drawShape(g, l2);
        }
    }    
}
